spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/AbstractServerHttpResponse.java

```diff
@@ -0,0 +1,306 @@
+        this.dataBufferFactory = dataBufferFactory;
+        this.headers = headers;
+        cookies = new LinkedMultiValueMap<>();
+    }
+
+    @Override
+    public final DataBufferFactory bufferFactory() {
+        return dataBufferFactory;
+    }
+
+    final State state() {
+        return state.get();
+    }
+
+    @Override
+    public HttpHeaders getHeaders() {
+        if (readOnlyHeaders != null) {
+            return readOnlyHeaders;
+        } else if (state.get() == State.COMMITTED) {
+            readOnlyHeaders = HttpHeaders.readOnlyHttpHeaders(headers);
+            return readOnlyHeaders;
+        } else {
+            return headers;
+        }
+    }
+
+    @Override
+    public MultiValueMap<String, ResponseCookie> getCookies() {
+        return state.get() == State.COMMITTED ? CollectionUtils.unmodifiableMultiValueMap(cookies) : cookies;
+    }
+
+    @Override
+    public void addCookie(ResponseCookie cookie) {
+        requireNonNull(cookie, "ResponseCookie must not be null");
+
+        if (state.get() == State.COMMITTED) {
+            throw new IllegalStateException(
+                    "Can't add the cookie " + cookie + "because the HTTP response has already been committed");
+        } else {
+            getCookies().add(cookie.getName(), cookie);
+        }
+    }
+
+    /**
+     * Return the underlying server response.
+     *
+     * <p><strong>Note:</strong> This is exposed mainly for internal framework
+     * use such as WebSocket upgrades in the spring-webflux module.
+     */
+    public abstract <T> T getNativeResponse();
+
+    @Override
+    public void beforeCommit(Supplier<? extends Mono<Void>> action) {
+        commitActions.add(action);
+    }
+
+    @Override
+    public boolean isCommitted() {
+        final State state = this.state.get();
+        return state != State.NEW && state != State.COMMIT_ACTION_FAILED;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public final Mono<Void> writeWith(Publisher<? extends DataBuffer> body) {
+        // For Mono we can avoid ChannelSendOperator and Reactor Netty is more optimized for Mono.
+        // We must resolve value first however, for a chance to handle potential error.
+        if (body instanceof Mono) {
+            return ((Mono<? extends DataBuffer>) body).flatMap(buffer -> {
+                touchDataBuffer(buffer);
+                final AtomicBoolean subscribed = new AtomicBoolean();
+                return doCommit(() -> {
+                    try {
+                        // Upstream code uses `Mono.fromCallable(()
+                        // -> buffer)` which does
+                        // nothing when `Subscription.cancel()` is
+                        // called. It can lead to
+                        // leaking of the buffer when an
+                        // HttpResponse is canceled.
+                        return writeWithInternal(Mono.just(buffer).doOnSubscribe(s -> subscribed.set(true))
```

Comment:
>  If I understand correctly this also affects upstream. 

It might be a problem only on our side. If we didn't recover the `CancelledSubscriptionException` exception,
https://github.com/line/armeria/blob/master/spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaServerHttpResponse.java#L93-L95
the buffer would be released by 
https://github.com/line/armeria/blob/96ee8e871b2834ed765e83d3fd317072fce346e4/spring/boot3-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/AbstractServerHttpResponse.java#L187

Anyway, it would be worth discussing it with upstream maintainers. Let me create an issue for it.